hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+  private final List<XceiverClientSpi.Validator> validators;
+  private final boolean verifyChecksum;
+  private final Function<BlockID, BlockLocationInfo> refreshFunction;
+  private final RetryPolicy retryPolicy;
+  private int retries;
+
+
+  public StreamBlockInput(
+      BlockID blockID, long length, Pipeline pipeline,
+      Token<OzoneBlockTokenIdentifier> token,
+      XceiverClientFactory xceiverClientFactory,
+      Function<BlockID, BlockLocationInfo> refreshFunction,
+      OzoneClientConfig config) throws IOException {
+    this.blockID = blockID;
+    LOG.debug("Initializing StreamBlockInput for block {}", blockID);
+    this.length = length;
+    setPipeline(pipeline);
+    tokenRef.set(token);
+    this.xceiverClientFactory = xceiverClientFactory;
+    this.validators = ContainerProtocolCalls.toValidatorList(
+        (request, response) -> validateBlock(response));
+    this.verifyChecksum = config.isChecksumVerify();
+    this.refreshFunction = refreshFunction;
+    this.retryPolicy =
+        HddsClientUtils.createRetryPolicy(config.getMaxReadRetryCount(),
+            TimeUnit.SECONDS.toMillis(config.getReadRetryInterval()));
+
+  }
+
+
+  public BlockID getBlockID() {
+    return blockID;
+  }
+
+  public long getLength() {
+    return length;
+  }
+
+  @Override
+  public synchronized long getPos() {
+    if (length == 0) {
+      return 0;
+    }
+    if (blockPosition >= 0) {
+      return blockPosition;
+    }
+
+    if (allocated && !buffersHaveData() && !dataRemainingInBlock()) {
+      Preconditions.checkState(
+          bufferOffsetWrtBlockDataData + buffersSize == length,
+          "EOF detected but not at the last byte of the chunk");
+      return length;
+    }
+    if (buffersHaveData()) {
+      // BufferOffset w.r.t to BlockData + BufferOffset w.r.t buffers +
+      // Position of current Buffer
+      return bufferOffsetWrtBlockDataData + bufferoffsets.get(bufferIndex) +
+          buffers.get(bufferIndex).position();
+    }
+    if (buffersAllocated()) {
+      return bufferOffsetWrtBlockDataData + buffersSize;
+    }
+    return 0;
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    checkOpen();
+
+    int dataout = EOF;
+    int len = 1;
+    int available;
+    while (len > 0) {
+      try {
+        acquireClient();
+        available = prepareRead(1);
+        retries = 0;
+      } catch (SCMSecurityException ex) {
+        throw ex;
+      } catch (StorageContainerException e) {
+        if (shouldRetryRead(e)) {
+          releaseClient();
+          refreshBlockInfo(e);
+          continue;
+        } else {
+          throw e;
+        }
+      } catch (IOException ioe) {
+        if (shouldRetryRead(ioe)) {
+          if (isConnectivityIssue(ioe)) {
+            releaseClient();
+            refreshBlockInfo(ioe);
+          } else {
+            releaseClient();
+          }
+          continue;
+        } else {
+          throw ioe;
+        }
+      }
```

Comment:
We can extract this out to a common function like:
```
int tryRead(int len) {
    try {
         acquireClient();
         available = prepareRead(len);
         retries = 0;
} catch (...) {...}
```
and call this in the subsequent read overrides to prevent duplication?